### PR TITLE
Add optional torch.export.ExportGraphSignature to ONNXProgram

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113686
* #112272
* #113495
* #112289
* __->__ #113477

When the ONNX model is exported from a torch.export.ExportedProgram, a
torch.export.ExportedGraphSignature is available with the specification
of the model inputs and outputs.

ExportedGraphSignature includes information such as the mapping between
the exported input/buffer/output ONNX name to the original pytorch input/buffer/output name.

It also specifies the kind of the input, such as user_input, parameter,
buffer or constant_tensor. Outputs kind can be user_output, loss_output,
buffer_mutation, etc

Such information can be useful to understand what the ONNX model expects
as inputs and how the output will look like when the ONNX input/output
differs from the original PyTorch input/output schema.

When the ONNX model is exported from a Callable or regular
torch.nn.MOdule, such information is not available and
ONNXProgram.model_signature will yield NOne

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
@BowenBao PTAL

File: torch/onnx/_internal/exporter.py [hunk 1]
```diff
@@ -638,6 +641,75 @@ def model_proto(self) -> onnx.ModelProto:  # type: ignore[name-defined]
...
+                    InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='arg7_1'), target=None)
+                ],
+                output_specs=[
+                    OutputSpec(kind=<OutputKind.BUFFER_MUTATION: 3>, arg=TensorArgument(name='add'), target='my_buffer2'),
+                    OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='_log_softmax'), target=None)
+                ]
+            )
+        """
+
+        return self._model_signature
```
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:711]: I feel this is only accurate up till the point for fx-graph in ExportedProgram. Further process inside our exporter that modifies the signature will make this outdated. Unless we apply similar io adapting to this too.

File: torch/onnx/_internal/exporter.py [hunk 2]
```diff
@@ -638,6 +641,75 @@ def model_proto(self) -> onnx.ModelProto:  # type: ignore[name-defined]
             raise self._export_exception
         return self._model_proto
 
+    @property
+    def model_signature(self) -> Optional[torch.export.ExportGraphSignature]:
+        """The model signature for the exported ONNX graph.
+
+        This information is relevant because ONNX specification often differs from PyTorch's,
+        resulting in a ONNX graph with input and output schema different from the actual PyTorch model implementation.
+        By using the model signature, the users can understand the inputs and outputs differences
```
Comment (BowenBao) [inline torch/onnx/_internal/exporter.py:650]: Due to prior comment, I'm not sure this is accurate.

File: torch/onnx/_internal/exporter.py (hunk 1 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/exporter.py:711]: when user model is `torch.export.ExportedProgram`, we don't have any pass that changes the signature.
The only pass applied to the resulting graph module is `fx_module = passes.InsertTypePromotion(diagnostic_context, fx_module).run()`.

Comment (BowenBao) [reply torch/onnx/_internal/exporter.py:711]: okay, this makes sense.

**Review by BowenBao — APPROVED** [review]

**Pull request closed without merging**